### ISSUE TITLE
[bug] garbagecollector not start successfully

### ISSUE THREAD

**Issue by shenlanse** [kubernetes/kubernetes#65894]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:

 /kind bug
> /kind feature

**What happened**:

garbage collector hang in `WaitForCacheSync`

https://github.com/kubernetes/kubernetes/blob/v1.9.2/pkg/controller/garbagecollector/garbagecollector.go#L140

```
	if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
		return
	}
```

Because `gc.dependencyGraphBuilder.IsSynced` return false continuely, controller not started.

**What you expected to happen**:

garbage collector works as expected.

**How to reproduce it (as minimally and precisely as possible)**:

start kube-controller-manager with commandLine params:

```
--leader-elect=false
```

**Anything else we need to know?**:

https://github.com/kubernetes/kubernetes/blob/v1.9.2/pkg/controller/garbagecollector/graph_builder.go#L278

```
func (gb *GraphBuilder) startMonitors() {
	gb.monitorLock.Lock()
	defer gb.monitorLock.Unlock()

	if gb.stopCh == nil {
		return
	}

	// we're waiting until after the informer start that happens once all the controllers are initialized.  This ensures
	// that they don't get unexpected events on their work queues.
	<-gb.informersStarted

	monitors := gb.monitors
	started := 0
	for _, monitor := range monitors {
		if monitor.stopCh == nil {
			monitor.stopCh = make(chan struct{})
			gb.sharedInformers.Start(gb.stopCh)
			go monitor.Run()
			started++
		}
	}
	glog.V(4).Infof("started %d new monitors, %d currently running", started, len(monitors))
}
```

StopCh is nil. It is passed from :
https://github.com/kubernetes/kubernetes/blob/v1.9.2/cmd/kube-controller-manager/app/controllermanager.go#L165

```
	if !s.LeaderElection.LeaderElect {
		run(nil)
		panic("unreachable")
	}
```

**I don't know why let `stopCh` couple with leaderElection. Whatever --leader-elect is set, stopCh should not be nil.**

**Environment**:
- Kubernetes version (use `kubectl version`):

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.2-36+bd3adc80d742eb-dirty", GitCommit:"bd3adc80d742eb3c260a38190baeb639f461e996", GitTreeState:"dirty", BuildDate:"2018-04-13T01:36:15Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.2-69+9faf4a6380d659-dirty", GitCommit:"9faf4a6380d659786df4556db45af1da04f9a8da", GitTreeState:"dirty", BuildDate:"2018-07-04T09:20:49Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by wgliang** [comment]
@shenlanse 
This issue had been fixed in PR(#57932).

**Comment by liggitt** [comment]
this was actually fixed in https://github.com/kubernetes/kubernetes/pull/57340 in [1.9.3](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#changelog-since-v192)


/close
